### ISSUE TITLE
Allow computing SHA256 digest ahead of time

### ISSUE THREAD

**Issue by anthonyrisinger** [CargoSense/ex_aws#272]
I am using something like this:

``` elixir
auth = ExAws.Auth.headers(:put, dest, :s3, config, headers, nil)
{:ok, client} = :hackney.request(:put, dest, auth ++ headers, :stream, [])
```

to open a **streaming** PUT request and avoid using the S3 Multipart API (which requires buffering 5MiB chunks and is too memory intensive for many concurrent uploads). The application is essentially just a proxy to S3, so there is no way for me to compute the SHA256 while directly streaming client-to-S3. However, if the client initiating the upload explicitly sends the digest, we should sign that instead.

I am proposing `ExAws.Auth.build_canonical_request` use the `x-amz-content-sha256` header as-is if `body = nil`: https://github.com/CargoSense/ex_aws/blob/master/lib/ex_aws/auth.ex#L82

So in short:
- Use `sha256(body)` if body is not nil
- Use `x-amz-content-sha256` if provided and body is nil
- Use `UNSIGNED-PAYLOAD` if body is nil and `x-amz-content-sha256` is missing

**Comment by benwilson512** [comment]
Interesting use case!

I should definitely let you set the sha manually.

Can you talk more about your upload scenario by the way?

**Comment by anthonyrisinger** [comment]
The thing I'm writing is an authenticated S3 proxy microservice for other components on our platform. It might need to handle many concurrent upload requests for a number of unrelated reasons. The service is just a docker container running on a cluster with other containers. Throughput of an individual blob is not important here, we just need to be able to get said blob to S3 in a reasonable time, and the whole blob will always originate from the same machine. The S3 Multipart API would be pretty useful for a distributed upload, but is a little overkill for a simple, albeit large, upload.

Typical way to handle large, proxied uploads is to read a chunk from the source/client and write it to the destination socket, limiting the proxy's memory consumption. Since this container is running with limited memory (say 128 or 256MiB), using the Multipart API would consume a fair amount of memory for 100 uploads (500MiB). We could spool the 5MiB chunks to disk, but this is an ephemeral container environment and writing to disk is not something you normally want to do, and ideally the rootfs is readonly or very constrained itself.

So ultimately, I'm just trying to do a simple PUT of a large blob from the perspective of a proxy that doesn't have access to the blob and does not want to buffer anything at all. I have it working right now by reading then writing 8KiB chunks.

**Comment by benwilson512** [comment]
If you're just trying to proxy uploads wouldn't it be easier to just ask it for a presigned upload URL and then have the clients do the upload?

**Comment by anthonyrisinger** [comment]
Hmm, I will look into that, I don't actually know a huge amount about what's possible or idiomatic with S3.

None of the other components know they are talking to S3, but maybe the interface could simply be "Use the service to checkout a URL, then PUT too it".

I do still think it's reasonable to use the SHA256 if provided, though I realize `ex_aws` does not support streaming without using the Multipart API.

**Comment by benwilson512** [comment]
Yeah I would say it's generally wiser to let the clients do the uploading. Amongst other reasons it balances the load according to the clients who are doing the most work. And yeah the interface is indeed as simple as "grab this url, push data to it". I'll look into supporting the given sha

**Comment by anthonyrisinger** [comment]
That makes sense. I didn't want the blob itself to pass through the proxy anyway, but I wasn't sure how else to do it and was not aware of the presigned URL option. Unless it turns out to be too annoying for some reason, I'm going to move down this path, especially if `ex_aws` already provides the facilities to generate said URLs.

**Comment by anthonyrisinger** [comment]
```elixir
config = %{scheme: "https://",
           host: "s3.amazonaws.com",
           region: "us-east-1",
           secret_access_key: "secret_access_key",
           access_key_id: "access_key_id"}
{:ok, url} = ExAws.S3.presigned_url(config, :put, "bucket", "path/object",
                                    query_params: [{"x-amz-meta-hai","there"}])
```

Thanks :-) I completely glazed over `ExAws.S3.presigned_url` and also did not realize the AWS V4 sig itself enables this functionality. NEAT. Well it was a good exercise to learn AWS and Elixir I suppose!

Feel free to close this ticket unless you want to support streaming uploads without using the Multipart API.
